django/core/handlers/asgi.py

```diff
@@ -223,6 +229,12 @@ async def process_request(request, send):
                     pass
         body_file.close()
 
+        try:
```

Comment:
I am now suspecting that this block should go _before_ the call to `body_file.close()`, in case it is still possible for something to try and read from the body file. However there are currently no tests that require this one way or the other (but the old behaviour would have been that the equivalent `request.close()` code would normally have run before `body_file.close()`)

Reply:
Confirmed: having had this patch rolled out for a while, we were getting a lot of exceptions saying `ValueError: I/O operation on a closed file` that were triggered by web crawler bots (but hadn't been happening under "normal" loads)

I modified the patch so that the new code block goes before the call to `body_file.close()` - and this seems to have fixed the errors.

I don't think there is anything in the tests at the moment that checks for this ordering, but I am confident now that the way it is now is the correct way around to have it.

Reply:
@nessita I will take a look, it should be possible

Reply:
This was a little tricky but hopefully `test_cancel_post_request_with_sync_processing` does this. 

If I swap the order of body_file.close() and the request finished block I get a test failure due to the same error being thrown in the test view at the point `request.body` is called.

This maybe seems like an artificial test but I think it is minimal to reproduce the issue I was finding - its quite common for code to be effectively wrapped in sync_to_async due to middleware

Reply:
Thank you! I'll review ASAP